File: Algorithm/Java/515. Find Largest Value in Each Tree Row.java
/**
* Author: WuLC
* Date:   2017-04-29 10:54:10
* Last modified by:   WuLC
* Last Modified time: 2017-04-29 10:54:31
* Email: liangchaowu5@gmail.com
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

// traverse the tree row by row
public class Solution 
{
    public List<Integer> largestValues(TreeNode root) 
    {
        if (root == null) return new ArrayList<Integer>();
        List<Integer> result = new ArrayList<Integer>();
        List<TreeNode> curr, next;
        curr = new ArrayList<TreeNode>();
        curr.add(root);
        
        while(curr.size()>0)
        {
            int maxVal = Integer.MIN_VALUE;
            next = new ArrayList<TreeNode>();
            for(int i=0; i<curr.size(); i++)
            {
                TreeNode tmp = curr.get(i);
                if (tmp.left != null) next.add(tmp.left);
                if (tmp.right != null) next.add(tmp.right);
                maxVal = Math.max(maxVal, tmp.val);
            }
            result.add(maxVal);
            curr = next;
        }
        return result;
    }
}

File: Algorithm/C++/515. Find Largest Value in Each Tree Row.cpp
/*
 * Created on Sat Jun 02 2018 21:27:38
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */


// traverse the tree level by level
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
 
class Solution 
{
    public:
        vector<int> largestValues(TreeNode* root) 
        {
            vector<int> result;
            if (root==NULL) return result;
            vector<TreeNode*> curr;
            curr.push_back(root);
            while(curr.size() > 0)
            {
                vector<TreeNode*> next;
                int tmp = curr[0]->val;
                for(int i=0; i < curr.size(); i++)
                {
                    if(curr[i]->left) next.push_back(curr[i]->left);
                    if(curr[i]->right) next.push_back(curr[i]->right);
                    tmp = std::max(tmp, curr[i]->val);
                }
                result.push_back(tmp);
                curr = next;
            }
            return result;
        }
};